templates/console/src/MyCompanyName.MyProjectName/Program.cs

```diff
@@ -28,19 +32,27 @@ public async static Task<int> Main(string[] args)
         {
             Log.Information("Starting console host.");
 
-            var builder = Host.CreateDefaultBuilder(args);
+            var builder = Host.CreateApplicationBuilder(args);
 
-            builder.ConfigureServices(services =>
-            {
-                services.AddHostedService<MyProjectNameHostedService>();
-                services.AddApplicationAsync<MyProjectNameModule>(options =>
-                {
-                    options.Services.ReplaceConfiguration(services.GetConfiguration());
-                });
-            }).AddAppSettingsSecretsJson().UseSerilog().UseAutofac().UseConsoleLifetime();
+            builder.Configuration.AddAppSettingsSecretsJson();
+            builder.Logging.ClearProviders().AddSerilog();
+
+            builder.Services.AddAutofacServiceProviderFactory();
+
+            builder.Services.AddSingleton<IHostLifetime, ConsoleLifetime>();
+
+            builder.Services.AddHostedService<MyProjectNameHostedService>();
+            await builder.Services.AddApplicationAsync<MyProjectNameModule>();
 
             var host = builder.Build();
-            await host.Services.GetRequiredService<IAbpApplicationWithExternalServiceProvider>().InitializeAsync(host.Services);
+
+            var application = host.Services.GetRequiredService<IAbpApplicationWithExternalServiceProvider>();
+            var applicationLifetime = host.Services.GetRequiredService<IHostApplicationLifetime>();
+
+            applicationLifetime.ApplicationStopping.Register(() => AsyncHelper.RunSync(() => application.ShutdownAsync()));
+            applicationLifetime.ApplicationStopped.Register(() => application.Dispose());
+
+            await application.InitializeAsync(host.Services);
```

Comment:
hi

We can consider merging these code into an extension method.

```
AddConsoleApplicationAsync<MyProjectNameModule, IHostedService>()
```

Reply:
Awesome. But `ApplicationBuilder` depends on `Microsoft.Extensions.Hosting`, do we need to create a new library for this or just create this extension method in the template project?


```csharp
public static class HostExtensions
{
    // Move to `Volo.Abp.Core` as a extension method.
    public static async Task InitializeAsync(this IHost host)
    {
        var application = host.Services.GetRequiredService<IAbpApplicationWithExternalServiceProvider>();
        var applicationLifetime = host.Services.GetRequiredService<IHostApplicationLifetime>();

        applicationLifetime.ApplicationStopping.Register(() => AsyncHelper.RunSync(() => application.ShutdownAsync()));
        applicationLifetime.ApplicationStopped.Register(() => application.Dispose());

        await application.InitializeAsync(host.Services);
    }

    // ?
    public static async Task AddConsoleApplicationAsync<TStartModule>(
        this HostApplicationBuilder builder,
        Action<AbpApplicationCreationOptions>? optionsAction = null)
        where TStartModule : AbpModule
    {
        builder.Services.AddSingleton<IHostLifetime, ConsoleLifetime>();

        await builder.Services.AddApplicationAsync<TStartModule>(optionsAction);
    }

    public static async Task AddConsoleApplicationAsync<TStartModule, THostedService>(
        this HostApplicationBuilder builder,
        Action<AbpApplicationCreationOptions>? optionsAction = null)
        where TStartModule : AbpModule
        where THostedService : IHostedService
    {
        // I think it is not necessary for console application.
        builder.Services.AddHostedService<IHostedService>();

        await builder.AddConsoleApplicationAsync<TStartModule>(optionsAction);
    }

    // Is it necessary?
    public static HostApplicationBuilder AddAutofac(this HostApplicationBuilder builder)
    {
        builder.Services.AddAutofacServiceProviderFactory();

        return builder;
    }
}
```

Reply:
Or instead of using `HostApplicationBuilder` directly, just use `IServiceCollection`?
```
public static class HostExtensions
{
    // Move to `Volo.Abp.Core` as a extension method.
    public static async Task InitializeAsync(this IHost host)
    {
        var application = host.Services.GetRequiredService<IAbpApplicationWithExternalServiceProvider>();
        var applicationLifetime = host.Services.GetRequiredService<IHostApplicationLifetime>();

        applicationLifetime.ApplicationStopping.Register(() => AsyncHelper.RunSync(() => application.ShutdownAsync()));
        applicationLifetime.ApplicationStopped.Register(() => application.Dispose());

        await application.InitializeAsync(host.Services);
    }
    
    // Move to Volo.Abp.Core `IServiceCollection` extensions
    public static async Task AddConsoleApplicationAsync<TStartModule>(
        this IServiceCollection services,
        Action<AbpApplicationCreationOptions>? optionsAction = null)
        where TStartModule : AbpModule
    {
        services.AddSingleton<IHostLifetime, ConsoleLifetime>();

        await services.AddApplicationAsync<TStartModule>(optionsAction);
    }

    public static async Task AddConsoleApplicationAsync<TStartModule, THostedService>(
        this IServiceCollection services,
        Action<AbpApplicationCreationOptions>? optionsAction = null)
        where TStartModule : AbpModule
        where THostedService : IHostedService
    {
        // I think it is not necessary for console application.
        services.AddHostedService<IHostedService>();

        await services.AddConsoleApplicationAsync<TStartModule>(optionsAction);
    }
}
```